### ISSUE TITLE
Feature Request : Automatic Sort?

### ISSUE THREAD

**Issue by barraIhsan** [tryashtar/nbt-studio#12]
NBT Studio seems to not sort a to z automatically
It kinda annoying when you open for example like the `level.dat`
![image](https://user-images.githubusercontent.com/57800056/115326106-f1560b80-a1b6-11eb-9c9b-73cebfd48afc.png)
So, when you want to sort, you have to click the object and click the sort button
![image](https://user-images.githubusercontent.com/57800056/115326160-0af75300-a1b7-11eb-99c4-92076110876c.png)
So please, maybe there is a setting that automatically sorts those, just like NBT Editor

**Comment by tryashtar** [comment]
Well, I tried implementing this, but it's a bit wonky. Let me know what you think:
[NbtStudio.zip](https://github.com/tryashtar/nbt-studio/files/6341235/NbtStudio.zip)

Mostly the challenges are how often to sort. I can make it sort every time you make a change, so all your new tags end up with their neighbors, and stuff like drag and drop would just ignore the position you selected for compounds. That might be potentially slow though, especially with bulk changes, since I currently don't really have a way to suppress multiple updates coming from the same source (it's something for me to look into later...)

The version I posted above just sorts whenever you open a file. Specifically, it sorts everything you have open at the top level, so if you import, it will re-sort already imported files, and it will miss stuff if you open a folder. Also it tends to mark every file you open as unsaved. If just sorting on read is what you want, I can probably make this process a little less hacky.

**Comment by barraIhsan** [comment]
> Well, I tried implementing this, but it's a bit wonky. Let me know what you think:
NbtStudio.zip

![image](https://user-images.githubusercontent.com/57800056/115352502-a00f4180-a1e1-11eb-9375-2ba13aa96695.png)
Also, why does the NBTEditor does this fine?

**Comment by tryashtar** [comment]
You have to click on the sort button to set which autosort you prefer. Sorry, I forgot to mention that.
NBTExplorer simply sorts the tags of the compound as it reads them from disk, which I'm not a fan of because I like to view and preserve the true order of tags. So at the very least, this feature requires more consideration from me because I want it to be configurable.
Also, testing in NBTExplorer just now, it looks like it has an issue where renaming a tag doesn't update its position until you add another tag beside it. That's the kind of thing I definitely want to avoid.

**Comment by barraIhsan** [comment]
Oh it work! Also it seems that when you open a file, it marked as * (not saved)
![image](https://user-images.githubusercontent.com/57800056/115354196-8ec73480-a1e3-11eb-9046-d54fda929ca4.png)
(opening the `level.dat_old` cause the `leve.dat` already sorted)

**Comment by tryashtar** [comment]
Yeah, I'll fix that later when I finish this feature. I'm curious, which of these two would you prefer?

* Compounds are always sorted. Adding new tags or renaming them puts them in order
* Files are sorted when loaded, but changes made after that can result in out-of-order tags

**Comment by barraIhsan** [comment]
> * Compounds are always sorted. Adding new tags or renaming them puts them in order

This

**Comment by tryashtar** [comment]
Alright good to know. About the settings, I thought I had that working, but I may have missed it.
Anyway, it's going to take some time before I finish this feature. Today's test was a good experiment to find out what work is required and where the pain points are. I do agree it's an important feature so I'll try my best

**Comment by barraIhsan** [comment]
Ok, thanks!

**Comment by annenveritas** [comment]
> * Compounds are always sorted. Adding new tags or renaming them puts them in order
> * Files are sorted when loaded, but changes made after that can result in out-of-order tags

I would vote against automatic sorting since it would make it harder to use the Studio for editing arbitrary NBT data: sometimes one wants their data just the order they've added it. A checkbox (or other toggleable setting) for automatic sorting would be nice though.

And by the way, there seems to be an issue with sorting of compounds that have nested compounds: their contents also get sorted, and that is irreversible - undoing restores just the immediate children's order, but the nested compounds' children stay sorted. The way I would prefer this, is sorting affecting just the immediate children of a compound chosen; though I understand that many users would probably prefer a deeper sorting so that they wouldn't have to sort manually all of the nested compounds (in case they want everything sorted), - so maybe a nice would be thing to give users a choice on how they want their sorting.

**Comment by tryashtar** [comment]
> I would vote against automatic sorting since it would make it harder to use the Studio for editing arbitrary NBT data: sometimes one wants their data just the order they've added it. A checkbox (or other toggleable setting) for automatic sorting would be nice though.

I agree 100%, I've already modified the NBT library to allow compound tags to have a consistent custom order because I think it's very important. Auto-sort will definitely be a default-off option.

The issues with undo are what's causing this to so long. If auto-sort counts as an undoable action (which it does currently because the codebase basically forces every action to be undoable), then every real action you take gums up the undo history, and undoing un-sorts. Which means either it's constantly unsorting and resorting, or you just can't undo ever, since it should re-sort right away. If auto-sort doesn't count as an undoable action, the program crashes when you try to undo because tags that used to be in one place are now somewhere else.

I really need a C# tutor to help me with my design...

**Comment by annenveritas** [comment]
> The issues with undo are what's causing this to take so long. If auto-sort counts as an undoable action (which it does currently because the codebase basically forces every action to be undoable), then every real action you take gums up the undo history, and undoing un-sorts. Which means either it's constantly unsorting and resorting, or you just can't undo ever, since it should re-sort right away. If auto-sort doesn't count as an undoable action, the program crashes when you try to undo because tags that used to be in one place are now somewhere else.

Hm, well, as a temporary workaround, maybe you could make the auto-sort actions somehow group with the last manual action, so that undoing once would first undo all of the auto-sorts (maybe with a temporarily internally disabled auto-sorting) and then it would undo that manual action, that followed by auto-sort again, but grouped with the action that was initially further in the undo history. This way redoing would be just repeating the manual action followed by an auto-sort.

Or maybe it could've been faster and easier just copying the state of the sorted compound just before the post-action auto-sort (which in this scenario wouldn't count as undoable), and restoring that state in case of an undo (thus avoiding all of the unsorting, though at the cost of some more RAM).

**Comment by tryashtar** [comment]
Combining it with the previous action isn't a bad idea, but at the moment I'm not sure whether the infrastructure allows for that. I also want to change my current bad event-driven NBT tags and nodes that run absurd amounts of code when anything changes to a kind of "mark dirty and synchronize later" approach. That will, I think, make autosort much easier to handle. But I'm really struggling with it.

**Comment by tryashtar** [comment]
Just a random update, I've been working on this rewrite continually since the end of July (and a bit earlier). It's been very slow and honestly quite stressful for me. The code didn't compile at all for around a month. Just last week, I fixed all the compiler errors and my pace improved, as I was able to test things. You may be disappointed to hear that so far, I am still working merely on *restoring* functionality I tore out in the process; I haven't even started working on auto-sort as a feature. I decided to push the [mark-dirty](https://github.com/tryashtar/nbt-studio/tree/mark-dirty) branch so anyone curious could see how my work was coming along.

The features I still need to restore include:
* Right-click context menus
* Refresh button
* Delete button
* SNBT buttons
* Cut/copy/paste
* Drag and drop
* Error messages when loading files in folders
* Regions and chunks

I am rather pleased with the state of the program infrastructure after the rewrite, though it took a lot of headaches to get it here, since I'm so amateurish. I appreciate your patience in the meantime; I hope that once I've finished restoring the existing feature set, implementing this and the remaining feature requests will be swift and elegant.
